crates/amaru-ledger/src/rules/transaction/scripts.rs

```diff
@@ -1,59 +1,115 @@
 use std::collections::BTreeSet;
 
 use amaru_kernel::{
-    get_provided_scripts, HasScriptHash, MintedWitnessSet, ScriptHash, TransactionInput,
+    get_provided_scripts, DatumOption, HasAddress, HasDatum, HasScriptHash, Hash, MintedWitnessSet,
+    OriginalHash, PseudoScript, ScriptHash, ScriptRefWithHash, TransactionInput,
 };
 use thiserror::Error;
 
 use crate::context::{UtxoSlice, WitnessSlice};
 
 #[derive(Debug, Error)]
 pub enum InvalidScripts {
-    #[error("missing required scripts: missing {0:?}")]
+    #[error("missing required scripts: missing [{}]",
+        .0.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+    )]
     MissingRequiredScripts(Vec<ScriptHash>),
-    #[error("extraneous script witnesses: extra {0:?}")]
+    #[error("extraneous script witnesses: extra [{}]",
+        .0.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+    )]
     ExtraneousScriptWitnesses(Vec<ScriptHash>),
+    #[error("unspendable inputs; no datums: [{}]",
+        .0
+        .iter()
+        .map(|input|
+            format!("{}#{}", input.transaction_id, input.index)
+        )
+        .collect::<Vec<_>>()
+        .join(", ")
+    )]
+    UnspendableInputsNoDatums(Vec<TransactionInput>),
+    #[error(
+        "missing required datums: missing [{}] provided [{}]",
+        missing.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+        provided.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+    )]
+    MissingRequiredDatums {
+        missing: Vec<Hash<32>>,
+        provided: BTreeSet<Hash<32>>,
+    },
+    #[error(
+        "extraneous supplemental datums: allowed: [{}] provided [{}]",
+        allowed.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+        provided.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
```

Comment:
```rust
.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", ")
```

is repeated 6 times in those 40 lines; perhaps consider moving it as a separate function?